src/libsyntax/parse/parser.rs

```diff
@@ -3210,7 +3228,23 @@ impl<'a> Parser<'a> {
             err.span_label(sp, "expected if condition here");
             return Err(err)
         }
-        let thn = self.parse_block()?;
+        let not_block = self.token != token::OpenDelim(token::Brace);
+        let fat_arrow_sp = if self.token == token::FatArrow {
+            Some(self.span)
+        } else {
+            None
+        };
+        let thn = self.parse_block().map_err(|mut err| {
+            if let Some(sp) = fat_arrow_sp {
+                // if cond => expr
+                err.span_suggestion(sp,
+                                    "only necessary in match arms, not before if blocks",
```

Comment:
Nit: I found this confusingly phrased. Maybe this?

```
"only necessary with `match`, not `if`"
```

I'm also aiming to avoid jargon like "arms" and "blocks", which users may not be familiar with.